### PR TITLE
feat: calculate linear progress

### PR DESCRIPTION
**Pull request by pedrolamas**
Klipper calculates progress by calculating `current_file_position / file_length`, but that can lead to progress starting on a non 0% value and the print complete before it reaches 100% (in my tests, I've seen it start at 46% and end at 97%)

The reason for this is that G-code files can have a lot of metadata on the start (thumbnails for example) and end of the file, thus causing the above problem.

As Moonraker reports the start byte and end byte of the G-code, we can use that together with the current file position and calculate a more correct progress value!

Having a "better" progress value will also lead to more accurate timing reports!

Before:

Animation

After:

Animation2

Signed-off-by: Pedro Lamas <pedrolamas@gmail.com>

### CODE REVIEW

**Review by matmen — APPROVED** [review]
LGTM, though I'd like to have some feedback from the others too

**Comment by pedrolamas** [comment]
Fair enough, will wait a few days for feedback from the rest of the team!

File: src/store/printer/getters.ts [hunk 1]
```diff
@@ -101,6 +101,19 @@ export const getters: GetterTree<PrinterState, RootState> = {
   },
 
   getPrintProgress: (state) => {
+    const { gcode_start_byte, gcode_end_byte, filename } = state.printer.current_file ?? {}
+    const { file_position } = state.printer.virtual_sdcard ?? {}
+
+    if (gcode_start_byte && gcode_end_byte && file_position && filename === state.printer.print_stats?.filename) {
```
Comment (eduncan911) [inline src/store/printer/getters.ts:107]: `filename === state.printer.print_stats?.filename`

With the `const` on line 104 defining `filename` in the scope of this method, and the way this `filename` is being defined inline creates an in-scope variable of, well, `filename`.

However, this variable is not used anywhere in this scope below, between lines 108 to 114.

Do we need this check?

**Review by eduncan911 — APPROVED** [review]
LGTM!  Just a coding question, nothing with the actual logic.

File: src/store/printer/getters.ts (hunk 1 above)
Comment (pedrolamas) [reply src/store/printer/getters.ts:107]: The `filename` on 104 is from `state.printer.current_file`, the check on 107 is with `state.printer.print_stats` (different parent objects).

This is a sanity check as the 2 parent objects can be updated at different times, making them out-of-sync.

**Pull request merged**